napari/components/viewer_model.py

```diff
@@ -995,6 +995,9 @@ def _add_layers_with_plugins(
             )
             # actually add the layer
             new = self._add_layer_from_data(*_data)
+            # store the source of the layer in metadata, if not present yet
+            if 'source' not in new.metadata:
+                new.metadata['source'] = filename
```

Comment:
```suggestion
            new.metadata.setdefault('source', filename)
```

good time to use [`dict.setdefault`](https://docs.python.org/3/library/stdtypes.html#dict.setdefault)!

Reply:
Great idea!